gadget/validate.go

```diff
@@ -230,31 +241,73 @@ func ensureRolesConsistency(roles map[string]*roleInstance, expectedSeed bool) e
 	return nil
 }
 
+func ensureRolesConsistencyClassicWithModes(roles map[string]*roleInstance) error {
+	if roles[SystemBoot] == nil || roles[SystemData] == nil {
+		return fmt.Errorf("system-boot and system-data roles are needed on classic")
+	}
+
+	// Make sure labels are as expected - save is optional
+	roleLabelToCheck := []roleLabel{roleLabelBoot, roleLabelData}
+	roleLabelOptional := []roleLabel{roleLabelSeed, roleLabelSave}
+	for _, rlLb := range roleLabelOptional {
+		if roles[rlLb.role] != nil {
+			roleLabelToCheck = append(roleLabelToCheck, rlLb)
+		}
+	}
+	if err := checkImplicitLabels(roles, roleLabelToCheck...); err != nil {
+		return err
+	}
+
+	// Check that boot/seed/save are in the same volume
+	bootVolName := roles[SystemBoot].volName
+	for _, rlLb := range roleLabelOptional {
+		if roles[rlLb.role] != nil && roles[rlLb.role].volName != bootVolName {
+			return fmt.Errorf("system-boot and %s are expected to share the same volume", rlLb.role)
+		}
+	}
+	return nil
+}
+
 func ensureSystemSaveRuleConsistency(roles map[string]*roleInstance) error {
 	if roles[SystemData] == nil || roles[SystemSeed] == nil {
 		// previous checks should stop reaching here
 		return fmt.Errorf("internal error: system-save requires system-seed and system-data structures")
 	}
-	if err := checkImplicitLabel(SystemSave, roles[SystemSave].s, ubuntuSaveLabel); err != nil {
+	if err := checkImplicitLabels(roles, roleLabelSave); err != nil {
 		return err
 	}
 	return nil
 }
 
 func checkSeedDataImplicitLabels(roles map[string]*roleInstance) error {
```

Comment:
this helper is a bit silly in the rewritten form, given that is called only once, maybe we should remove it and just embed the implementation call in its place?

Reply:
Good point, removed now